ENH: tilde expansion for write functions, #11438

aechase opened the pull request:
closes #11438 
Added calls to `_expand_user()` in multiple Formatter classes so that writer functions, such as `to_csv()` and `to_html()`, can perform tilde expansion on file paths beginning `~/`. This makes the writer functions consistent with the reader functions, such as `read_csv()`. I also changed the `__init__()` for `CSVFormatter` to conform to the pep8 width limit.

jreback commented:
tests pls

jreback commented on pandas/core/format.py:
```diff
@@ -68,12 +68,14 @@
                  na_rep='NaN', footer=True):
         self.categorical = categorical
         self.buf = buf if buf is not None else StringIO(u(""))
+        self.buf = _expand_user(self.buf)
```
doesn't belong here

jreback commented on pandas/core/format.py:
```diff
@@ -130,6 +132,7 @@ def __init__(self, series, buf=None, length=True, header=True,
                  max_rows=None):
         self.series = series
         self.buf = buf if buf is not None else StringIO()
+        self.buf = _expand_user(self.buf)
```
eoesn't belong here

jreback replied:
here ok too

jreback commented on pandas/core/format.py:
```diff
@@ -923,6 +927,7 @@ def write_tr(self, line, indent=0, indent_delta=4, header=False,
     def write_result(self, buf):
         indent = 0
         frame = self.frame
+        buf = _expand_user(buf)
```
doesn't belong here

jreback commented on pandas/core/format.py:
```diff
@@ -344,6 +347,7 @@ def __init__(self, frame, buf=None, columns=None, col_space=None,
                  max_cols=None, show_dimensions=False, **kwds):
         self.frame = frame
         self.buf = buf if buf is not None else StringIO()
+        self.buf = _expand_user(self.buf)
```
this is I think only place you need it (covers both CSV and HTML)

jreback commented:
actually on 2nd thought.

I think you these should ONLY be used in ``core/frame.py/DataFrame.to_csv and to_html``

I don't really like this here, as all of the other IO functions are localized in ``pandas/io/*`` but that's how it is for now.

aechase commented:
I had the same thought, but then I assumed that would cause problems for Series.to_csv(). But, having now checked the code, I see that they are the same function, with each Series being temporarily converted to a DataFrame.

Should I put the _expand_user() calls in frame.py?

Regarding the tests: let me start out by saying that I am oh-so-close to being out of my depth here. The current suite seems to run OK (various DeprecationWarnings and RuntimeWarnings notwithstanding), but I'm a bit unclear as to how (or whether) I should create a new test function. That function would need to write and read some temporary files, but it looks like it would be difficult, if not impossible, to use testing.ensure_clean() to generate expanded and non-expanded file paths.

jreback commented:
you might have a look at these: https://github.com/pydata/pandas/blob/master/pandas/io/tests/test_common.py

yeh this might be quite tricky to test

aechase committed:
ENH: tilde expansion for write functions

Closes GH11438

Enables DataFrame.to_csv(), .to_html() and .to_latex() to perform tilde expansion on file paths.

aechase commented:
@jreback looks like I might have caused some trouble editing the most recent version of the release notes file...

jreback commented:
merged via https://github.com/pydata/pandas/commit/ea758cd2467f3b9dbfe4aced1ce6666145c502c6

thanks!
